ext/MitsubaLoader/CMitsubaMaterialCompilerFrontend.cpp

```diff
@@ -0,0 +1,597 @@
+        }
+        break;
+        case CElementBSDF::COATING:
+        case CElementBSDF::ROUGHCOATING:
+        {
+            nextSym = ir->allocNode<IR::CCoatingBSDFNode>();
+            nextSym->children.count = 1u;
+
+            const float eta = current->dielectric.intIOR/current->dielectric.extIOR;
+
+            auto* node = static_cast<IR::CCoatingBSDFNode*>(nextSym);
+            node->thickness = current->coating.thickness;
+            getSpectrumOrTexture(current->coating.sigmaA, node->sigmaA);
+            node->eta = IR::INode::color_t(eta);
+            node->shadowing = IR::CCoatingBSDFNode::EST_SMITH;
+            if (currType == CElementBSDF::ROUGHCOATING)
+            {
+                node->ndf = ndfMap[current->coating.distribution];
+                getFloatOrTexture(current->coating.alphaU, node->alpha_u);
+                if (node->ndf == IR::CMicrofacetSpecularBSDFNode::ENDF_ASHIKHMIN_SHIRLEY)
+                    getFloatOrTexture(current->coating.alphaV, node->alpha_v);
+                else
+                    node->alpha_v = node->alpha_u;
+            }
+            else
+            {
+                node->setSmooth();
+            }
+            bsdfQ.push(current->coating.bsdf[0]);
+        }
+        break;
+        case CElementBSDF::BLEND_BSDF:
+        {
+            nextSym = ir->allocNode<IR::CBSDFBlendNode>();
+            nextSym->children.count = 2u;
+
+            auto* node = static_cast<IR::CBSDFBlendNode*>(nextSym);
+            getFloatOrTexture(current->blendbsdf.weight, node->weight);
+
+            bsdfQ.push(current->blendbsdf.bsdf[1]);
+            bsdfQ.push(current->blendbsdf.bsdf[0]);
+        }
+        break;
+        case CElementBSDF::MIXTURE_BSDF:
+        {
+            nextSym = ir->allocNode<IR::CBSDFMixNode>();
+            auto* node = static_cast<IR::CBSDFMixNode*>(nextSym);
+            const size_t cnt = current->mixturebsdf.childCount;
+            nextSym->children.count = cnt;
+            for (int32_t i = cnt-1u; i >= 0; --i)
+                node->weights[i] = current->mixturebsdf.weights[i];
+
+            for (int32_t i = cnt-1u; i >= 0; --i)
+                bsdfQ.push(current->mixturebsdf.bsdf[i]);
+        }
+        break;
+        }
+
+        IR::INode* newParent = nullptr;
+        if (nextSym->children)
+            nodeQ.push(nextSym);
+        if (!--childrenCountdown && !nodeQ.empty())
+        {
+            newParent = nodeQ.front();
+            nodeQ.pop();
+        }
+
+        parent->children[childrenCountdown] = nextSym;
+        if (newParent)
+        {
+            parent = newParent;
+            childrenCountdown = parent->children.count;
+        }
+    }
+
+    static_cast<IR::CMaterialNode*>(root)->opacity = opacity;
+    static_cast<IR::CMaterialNode*>(root)->thin = thin;
+
+    IR::INode* surfParent = root;
+    if (surfParent->children[0]->symbol == IR::INode::ES_GEOM_MODIFIER)
+        surfParent = surfParent->children[0];
```

Comment:
I don't get this.

Reply:
if there's geom modifier (i.e. bumpmap) at the top of AST, it becomes parent of surfaces in IR. Actually i dont like this and wanted to change but i forgot and left as is

Reply:
I think its related to this
https://github.com/buildaworldnet/IrrlichtBAW/pull/424#discussion_r509194399

Which is basically why do you need a special root node type anyway?

Reply:
https://github.com/buildaworldnet/IrrlichtBAW/pull/424#discussion_r514558166